### PR TITLE
Feature/inner kore improvements

### PR DESCRIPTION
**Pull request by cos**
**please review**, @dwightguth, @osa1 
- changed the underlying representation of KList to a linked list
- made most other KORE classes be treatable as a KList (at least from Scala)
- changed the signature of the Transformers to match Java 8 lambdas
- added many tests

### CODE REVIEW

**Comment by cos** [comment]
Jenkins, test this

**Comment by osa1** [comment]
lgtm.

**Comment by dwightguth** [comment]
KList is supposed to be associative, isn't it? Why does this data structure seem to be a cons list then?

**Comment by grosu** [comment]
I just to strengthen that: KList is associative!  Also, in KORE, we want the "assoc" attribute to make it clear that any K label having it is associative, that is, 

```
syntax KLabel ::= "abc" [assoc]
```

means that I can freely write `abc(.KList)` or `abc(K1)` or `abc(K1,K2,K3)` in rules, and do associative matching on them.  Moreover, if the operation is also `comm`, then we will want both associative and commutative matching on its arguments (given as a KList).  Of course, this has nothing to do with parsing right now, but let's keep it in mind for later.

**Comment by cos** [comment]
Could somebody explain how the underlying implementation influences the way the data structures behave during matching? Which would be the best data structure type and why?

**Comment by grosu** [comment]
Yes, we should not worry about execution at parsing stage.  But, you will still need to know something about it in order to parse rules relying on associative matching.  For example, if `abc` is some vararg multiplication operation, then I want to write the rule:

```
rule abc(L1:Klist, 0, L2:Klist) => 0
```

How do you store it in KORE if KList is a cons list?

**Comment by cos** [comment]
Right now it is stored as:

```
KApply(KLabel("abc"), KList(KVariable("L1"), KToken(Int, 0), KVariable("L2")))
```

The `KList` above generates internally `ConsKList(Kvariable("L1"), ConsKList(KToken(Int, 0), ConsKList(KVariable("L2"), EmptyKList)))`, but this has nothing to do with any of the operations this entire data structure responds to. It has exactly the same interface as it would have it would be backed by a different data structure. The only difference is in the performance characteristics. For the common operation in the compiler (i.e., traversal), the linked list is likely the fastest choice. It is only slower on indexed access, but do we ever do that?

I am not arguing that it is irrelevant that `KList` is associative. I just think it is irrelevant in the choice of underlying data structure. Whether we use a an array-backed list or a cons list to hold the `KList` doesn't seem to make any difference. Both (and likely any other) will also require some form of normalization to make sure we have good "equals". Backends may make other choices for speeding up matching.

In any case, which data structure do you suggest as a better choice?

**Comment by dwightguth** [comment]
Maybe I'm misunderstanding, but doesn't a cons list have inefficient append
operations?

**Comment by cos** [comment]
@dwightguth, not necessarily. Appending is just linking the tail to the new list/element. The operation itself is O(1). If you don't know which is last element, it can become O(n). It is O(n) right now, but we can easily make it O(1) by remembering the tail. We would pay some price in terms of memory then. The price is only worth it if append is a very common operation in the compiler. I don't think the discussion about optimizations is worth it now, as long as the underlying representation doesn't leak into the interfaces -- and it doesn't.

**Comment by dwightguth** [comment]
Okay. Can you show me where in this diff the interface for this KList is
then?

**Comment by cos** [comment]
Simply consider only the `KList` abstract class. The interface is gathered from several traits. In a nutshell, it has a `KList(K... elements)` constructor and head/tail/apply/find/map/reduce/fold etc. operations (just Scala `List`). 

Again, if we decide to use something else underneath, none of these will disappear, they will just have different performance characteristics.

**Comment by cos** [comment]
Once they settle a bit (and we're close), I will document how to use the classes from Java on the wiki. The operations are very homogenous, so I think a few pages on wiki coupled with the `kore.k` syntax document will be enough to know how to use everything.

File: kore/src/main/scala/org/kframework/kore/KList.scala [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+  self: This =>
+  val klist: KList
+
+  override def head = klist.head
+  override def tail = copy(klist.tail)
+  override def isEmpty = klist.isEmpty
+}
+
+/**
+ * Should be extended by companion objects of classes extending KListLike
```
Comment (dwightguth) [inline kore/src/main/scala/org/kframework/kore/KList.scala:68]: companion object? can you provide an example of what that means?

**Comment by cos** [comment]
Scala doesn't have the concept of static members (fields or methods). Instead, it uses a singleton object with the same name of as the class. This singleton is called the "companion object" of the class.

The `CanBuildKListLike` trait is used for defining all the Scala implicits needed for making operations on `KListLike` collections return the same collection. E.g., in Scala: `KSequence(...).map(x => x)` has the type `KSequence`. There is no special `map` operation defined in `KSequence`. Everything is done via implicits defined by the companion object `KSequence` through extending `CanBuildKListLike` -- no extra code in either the `KSequence` class or its companion object.

For Java, as it doesn't have any similar mechanism, I had to add some extra code to achieve the same interface for `map`, and will probably extends to a few more (e.g., `filter`). 

I intend the "go to" class for extension to be `KCollection`, and it does not require the companion object when extended from Java. Again, the companion object only makes the Scala interfaces a bit nicer.

**Comment by dwightguth** [comment]
okay. this diff seems fine then.

**Comment by cos** [comment]
I think I can make the code needed for the Java `map` interface a bit smaller. Will merge this and will make a PR for that a bit later.

**Pull request merged**